### PR TITLE
Replaced generic ValueError with custom subclass when reporting unexpected resource status

### PR DESCRIPTION
**Pull request by arodiss**
*Issue #, if available:*
None

*Description of changes:*
Replaced generic ValueError with custom subclass when reporting unexpected resource status

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#commit-message-guidlines)
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=56ddae45-e536-40fd-ab14-5c8f30275e1b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=fb8896e0-82a6-4f49-a4ce-6c2a404b47fc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by mvsusp** [comment]
Hello @arodiss, thanks for the contribution. Any special reasoning behinds creating a new special exception? I would tend to just use `ValueError` here otherwise.

**Comment by arodiss** [comment]
@mvsusp 
I want to catch this specific situation in my client script, and to distinguish between "batch transformation status is failed because some records are malformed" and "ValueError was raised by one of the 10000 possible sources" I have to parse error message.
Generally, I think that specific exception classes are strictly preferable over built-ins - they *allow* something for a client but do not *demand* anything from him.
Here my custom exception is inherited from the `ValueError`, so if somebody is not interested in specific handling - his code will work just as it did before

File: src/sagemaker/session.py [hunk 1]
```diff
@@ -712,8 +713,11 @@ def wait_for_model_package(self, model_package_name, poll=5):
...
-            raise ValueError('Error creating model package {}: {} Reason: {}'.format(
-                model_package_name, status, reason))
+            raise UnexpectedStatusException(
+                message='Error creating model package {}: {} Reason: {}'.format(model_package_name, status, reason),
+                allowed_statuses=['Completed'],
+                actual_status=status
+            )
         return desc
 
     def create_endpoint_config(self, name, model_name, initial_instance_count, instance_type,
```
Comment (mvsusp) [inline src/sagemaker/session.py]: Please include unit tests, example https://github.com/aws/sagemaker-python-sdk/blob/2763f9ac16cae778905dddbaa69ea63f4ae39adc/tests/unit/test_create_deploy_entities.py#L61

File: src/sagemaker/session.py [hunk 2]
```diff
@@ -921,7 +925,11 @@ def _check_job_status(self, job, desc, status_key_name):
         if status != 'Completed' and status != 'Stopped':
             reason = desc.get('FailureReason', '(No reason provided)')
             job_type = status_key_name.replace('JobStatus', ' job')
-            raise ValueError('Error for {} {}: {} Reason: {}'.format(job_type, job, status, reason))
+            raise UnexpectedStatusException(
+                message='Error for {} {}: {} Reason: {}'.format(job_type, job, status, reason),
+                allowed_statuses=['Completed', 'Stopped'],
+                actual_status=status
+            )
```
Comment (mvsusp) [inline src/sagemaker/session.py]: We need unit tests here as well.

File: src/sagemaker/session.py [hunk 3]
```diff
@@ -938,7 +946,11 @@ def wait_for_endpoint(self, endpoint, poll=5):
 
         if status != 'InService':
             reason = desc.get('FailureReason', None)
-            raise ValueError('Error hosting endpoint {}: {} Reason: {}'.format(endpoint, status, reason))
+            raise UnexpectedStatusException(
+                message='Error hosting endpoint {}: {} Reason: {}'.format(endpoint, status, reason),
+                allowed_statuses=['InService'],
+                actual_status=status
+            )
         return desc
```
Comment (mvsusp) [inline src/sagemaker/session.py]: Please include unit tests.

File: src/sagemaker/session.py [hunk 4]
```diff
@@ -30,6 +30,7 @@
 from sagemaker import vpc_utils
 from sagemaker.user_agent import prepend_user_agent
 from sagemaker.utils import name_from_image, secondary_training_status_changed, secondary_training_status_message
+from sagemaker.exceptions import UnexpectedStatusException
```
Comment (mvsusp) [inline src/sagemaker/session.py]: Let's try to follow Google Python Style guide, http://google.github.io/styleguide/pyguide.html#224-decision

```python
from sagemaker import exceptions
```

**Comment by mvsusp** [comment]
Agreed that is helpful in your use case. Thanks for your contribution!

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=8c2e726d-e47b-489e-a6b9-711fade4b233%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=17df8b32-b138-4d06-abc9-eeaed64b7abe%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: src/sagemaker/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (arodiss) [reply src/sagemaker/exceptions.py]: added

File: src/sagemaker/session.py (hunk 1 above)
Comment (arodiss) [reply src/sagemaker/session.py]: included

File: src/sagemaker/session.py (hunk 2 above)
Comment (arodiss) [reply src/sagemaker/session.py]: done

File: src/sagemaker/session.py (hunk 3 above)
Comment (arodiss) [reply src/sagemaker/session.py]: done

File: src/sagemaker/session.py (hunk 4 above)
Comment (arodiss) [reply src/sagemaker/session.py]: updated.
I have to mention that other imports in the same file are quite far from the guide

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=46c7c49a-36df-43ef-8457-4bce01faa825%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by arodiss** [comment]
@mvsusp 
All done.
Test failures are same as on the master

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=2386471e-adac-4048-a440-cf4901e2fe90%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://hno1ypgnog.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=fb9d19d1-3116-4e22-92dc-7a35dcbf5380%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://rp2is7byfb.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=afa5f033-9d33-4e59-8e04-6a3f6ea11c58%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://mdk06ma93j.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4444e4dc-ed10-4af1-af36-26ee1d4ee4bf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://rp2is7byfb.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=8da433bf-aa80-4da7-b3d7-3b068bccaa34%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://hno1ypgnog.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=77390c63-949c-4de2-80e9-f812e2698573%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://mdk06ma93j.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=1aa0a6e5-640f-45fc-b312-3e903f35df58%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4bc1bd3b-317a-48ae-a2b5-bbf7253c0cd4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by arodiss** [comment]
Moved to #919

**Pull request closed without merging**